### PR TITLE
saves SCM widget input on window reload

### LINKED ISSUE #41232: SCM: Restore commit message on reload/reopen
Sometimes I'm half-way done typing a commit message and I end up doing something else, then for some reason or other reload/reopen the editor, and what I typed is gone. Obviously not the end of the world, but even the search keeps its state when reloading/reopening. I think the extension search is the only other thing doesn't.

### PR DESCRIPTION
**Pull request by meganrogge**
Saves the SCM widget input on window reload

This PR fixes #41232

### CODE REVIEW

**Review by joaomoreno — CHANGES_REQUESTED** [review]
@meganrogge Note that you can open multiple SCM repositories in a single window. Try opening a git repo and then `File > Add Folder to Workspace` to add another git repo. Both repositories will be open in the SCM view. The current proposal will mix up their stored input values since a single key is being used: `scm.text`. This should definitely be scoped per repository.

Also note that the input value should be synchronized between VS Code and the extension API, where SCM providers can access that value (see further down `Keep API in sync with model,`). This solution seems to set the value in the text editor on startup, but fails to propagate that value back to the API.

I wonder if this should be done at a deeper model layer, instead of inside the SCMInputWidget. Perhaps the [`SCMInput` model](https://github.com/microsoft/vscode/blob/0bf0ecf9de78d1c4b87342395be4b0bb2482c556/src/vs/workbench/contrib/scm/common/scmService.ts#L21-L26) should have that _memory_ instead. What do you think?

File: src/vs/workbench/contrib/scm/common/scm.ts [hunk 1]
```diff
@@ -84,6 +85,8 @@ export interface IInputValidator {
 }
 
 export interface ISCMInput {
+	store(value: string): void;
+	storedValue(): string | undefined;
```
Comment (joaomoreno) [inline src/vs/workbench/contrib/scm/common/scm.ts]: Thinking ahead (https://github.com/microsoft/vscode/issues/26952)... would a different interface here make more sense? What do you think about having a read-only array of values, acting as a history? Something like:

```ts
export interface ISCMInput {
  readonly history: readonly string[];
}
```

The `ISCMInput` interface should really only expose a read-only value/history. No other entity other than `SCMInput` should be allowed to modify it. So, that history would be updated simply by setting new `value` on `SCMInput`, ie in the setter of `SCMInput.value`.

**Review by joaomoreno — CHANGES_REQUESTED** [review]
More comments

File: src/vs/workbench/contrib/scm/common/scmService.ts [hunk 2]
```diff
@@ -71,7 +72,18 @@ class SCMInput implements ISCMInput {
 	private readonly _onDidChangeValidateInput = new Emitter<void>();
 	readonly onDidChangeValidateInput: Event<void> = this._onDidChangeValidateInput.event;
 
-	constructor(readonly repository: ISCMRepository) { }
+	constructor(readonly repository: ISCMRepository, @IStorageService memory: IStorageService) {
+		this.memory = memory;
+	}
+	memory: IStorageService;
```
[the remark this answers is not in the document]
Comment (meganrogge) [reply src/vs/workbench/contrib/scm/common/scmService.ts]: Both points make sense. I usually abide by that naming convention, but thought you meant in the above that _memory_ should be the name.

File: src/vs/workbench/contrib/scm/common/scm.ts (hunk 1 above)
Comment (meganrogge) [reply src/vs/workbench/contrib/scm/common/scm.ts]: I like that idea and will explore it today.

File: src/vs/workbench/contrib/scm/common/scmService.ts [hunk 3]
```diff
@@ -71,7 +73,23 @@ class SCMInput implements ISCMInput {
...
 
-	constructor(readonly repository: ISCMRepository) { }
+	constructor(readonly repository: ISCMRepository, @IStorageService private storageService: IStorageService) {
+		this._value = this.storedValue() ?? '';
+	}
+
+	store(value: string) {
+		let root = this.repository.provider.rootUri;
+		if (root) {
+			this.storageService.store(root.path, value, StorageScope.WORKSPACE);
```
Comment (joaomoreno) [inline src/vs/workbench/contrib/scm/common/scmService.ts:83]: Using `root.path` isn't a great idea for a couple reasons:
- the storage service keys are shared across the entire workbench, so we should prefix it with something related to SCM
- there can be two different SCM providers with the same root path, we should take `provider.label` also into account

How about:

```suggestion
			const key = `scm/input:${this.repository.provider.label}:${root.path}`;
			this.storageService.store(key, value, StorageScope.WORKSPACE);
```

**Review by joaomoreno — CHANGES_REQUESTED** [review]
Getting there!

**Review by joaomoreno — COMMENTED** [review]
@meganrogge It would have been better to get to an approved state before merging this in, though I do admire the passion, no harm done. I've given it a final test pass, namely concerning git commit templates, which still seem to work OK with this. I've also pushed an extra commit which addresses my current issues with the code: https://github.com/microsoft/vscode/commit/bd074c39b4b02c01d6d28af48f699cf693c0359e

cc @mjbvz

**Pull request merged**